File: typescript/rover.ts
export enum Command {
  F,
  B,
  L,
  R,
}
export enum Direction {
  N,
  E,
  S,
  W,
}
export interface Rover {
  location: [number, number];
  direction: Direction;
}

export const initial: Rover = { location: [0, 0], direction: Direction.N };

function moveForward(r: Rover): Rover {
  let [x, y] = r.location
  switch (r.direction) {
    case Direction.N: return { direction: r.direction, location: [x + 1, y] }
    case Direction.E: return { direction: r.direction, location: [x, y + 1] }
    case Direction.S: return { direction: r.direction, location: [x - 1, y] }
    case Direction.W: return { direction: r.direction, location: [x, y - 1] }
  }
}

function moveBackward(r: Rover): Rover {
  let [x, y] = r.location
  switch (r.direction) {
    case Direction.N: return { direction: r.direction, location: [x - 1, y] }
    case Direction.E: return { direction: r.direction, location: [x, y - 1] }
    case Direction.S: return { direction: r.direction, location: [x + 1, y] }
    case Direction.W: return { direction: r.direction, location: [x, y + 1] }
  }
}

function turnLeft(r: Rover): Rover {
  switch (r.direction) {
    case Direction.N: return { direction: Direction.W, location: r.location }
    case Direction.E: return { direction: Direction.N, location: r.location }
    case Direction.S: return { direction: Direction.E, location: r.location }
    case Direction.W: return { direction: Direction.S, location: r.location }
  }
}

function turnRight(r: Rover): Rover {
  switch (r.direction) {
    case Direction.N: return { direction: Direction.E, location: r.location }
    case Direction.E: return { direction: Direction.S, location: r.location }
    case Direction.S: return { direction: Direction.W, location: r.location }
    case Direction.W: return { direction: Direction.N, location: r.location }
  }
}

export function executeCommand(r: Rover, c: Command): Rover {
  switch (c) {
    case Command.F: return moveForward(r)
    case Command.B: return moveBackward(r)
    case Command.L: return turnLeft(r)
    case Command.R: return turnRight(r)
  }
}

export function executeCommands(r: Rover, c: Command[]): Rover {
  return c.reduce(executeCommand, r);
}

export function equal(r1: Rover, r2: Rover) {
  return r1.direction === r2.direction && r1.location === r2.location;
}

File: property-based-testing-workshop-clj/src/rover.clj
(ns rover
  (:require [clojure.test.check :as tc]
            [clojure.test.check.generators :as gen]
            [clojure.test.check.properties :as prop]))

(def initial {:direction :N :location [0 0]})

(def directions #{:N :S :E :W})
(def commands #{:F :B :L :R})


(defn move-forward [{direction :direction [x y] :location}]
  (let [new-location (condp = direction
                       :N [(inc x) y]
                       :S [(dec x) y]
                       :E [x (inc y)]
                       :W [x (dec y)])]
    {:direction direction :location new-location}))

(defn move-backward [{direction :direction [x y] :location}]
  (let [new-location (condp = direction
                       :N [(dec x) y]
                       :S [(inc x) y]
                       :E [x (dec y)]
                       :W [x (inc y)])]
    {:direction direction :location new-location}))

(defn rotate-left [{direction :direction location :location}]
  (let [new-direction (condp = direction
                       :N :W
                       :S :E
                       :E :N
                       :W :S)]
    {:direction new-direction :location location}))

(defn rotate-right [{direction :direction location :location}]
  (let [new-direction (condp = direction
                        :N :E
                        :S :W
                        :E :S
                        :W :N)]
    {:direction new-direction :location location}))

(def commands-by-id {:F move-forward
                     :B move-backward
                     :L rotate-left
                     :R rotate-right})

(defn execute-commands [rover cmds]
  (reduce (fn [r cmd] (cmd r))
    rover
    (map commands-by-id cmds)))

(comment
  ;; rovers are specified by a direction and a location
  (def a-rover {:direction :N :location [1 0]})

  ;; forms in a comment block can be easily sent to the repl
  (= (move-forward initial) {:direction :N :location [1 0]})
  (= (execute-commands initial (list :F :B :L :L :L :L)) initial)

  ,,,)

;; Let the testing commence!
(comment
  ;; use this to generate samples for easy viewing!
  (gen/sample (gen/elements commands))
  ,,,)

(def any-command-changes-the-rover
  (prop/for-all [r (gen/elements commands)]
    (not= initial (execute-commands initial [r]))))

(tc/quick-check 100 any-command-changes-the-rover)
